packages/expo-modules-core/ios/Swift/Logging/PersistentFileLog.swift

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+import Foundation
+
+public typealias PersistentFileLogFilter = (_: String) -> Bool
+public typealias PersistentFileLogCompletionHandler = (_:Error?) -> Void
+
+/**
+ * A thread-safe class for reading and writing line-separated strings to a flat file.
+ * The main use case is for logging specific errors or events, and ensuring that the
+ * logs persist across application crashes and restarts (for example, OSLogReader can
+ * only read system logs for the current process, and cannot access anything logged
+ * before the current process started).
+ *
+ * All write access to the file goes through asynchronous public methods managed by a
+ * serial dispatch queue.
+ *
+ * The dispatch queue is global, to ensure that multiple instances accessing the same
+ * file will have thread-safe access.
+ *
+ * The only operations supported are
+ * - Read the file (synchronous)
+ * - Append one or more entries to the file
+ * - Filter the file (only retain entries that pass the filter check)
+ * - Clear the file (remove all entries)
+ *
+ */
+public class PersistentFileLog {
+  private static let EXPO_UPDATES_LOG_QUEUE_LABEL = "dev.expo.updates.logging"
+  private static let serialQueue = DispatchQueue(label: EXPO_UPDATES_LOG_QUEUE_LABEL)
+
+  private let category: String
+  private let filePath: String
+
+  public init(category: String) {
+    self.category = category
+    let fileName = "\(PersistentFileLog.EXPO_UPDATES_LOG_QUEUE_LABEL).\(category).txt"
+    self.filePath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first?.appendingPathComponent(fileName).path ?? ""
```

Comment:
we probably shouldn't fall back to an empty string for the file path?

Reply:
That is not something that should ever happen, and if it did, everything would throw exceptions and just not work

Reply:
Yep, we should probably throw an exception here instead of in the other methods so that the cause is attributed correctly. If we wait for the other methods to throw, it'll require extra debugging to figure out that it failed here.

Reply:
@tsapeta has advised strongly against force-unwrapping in Swift. I'll take another look at this

Reply:
If it's going to crash anyways at some point, it's better to crash here instead since this is the cause.

Reply:
Ok done